### PR TITLE
Add train & submission

### PR DESCRIPTION
**Pull request by trushkin-ilya**
- add training
- add submission

### CODE REVIEW

**Comment by grib0ed0v** [comment]
It looks like you forgot to add **.ipynb_checkpoints** to .gitignore file

**Comment by grib0ed0v** [comment]
What is the purpose of having 'Untitled.ipynb'?

File: src/train.py [hunk 1]
```diff
@@ -1,20 +1,66 @@
...
+from torch import optim, nn
+from torchvision import models
+import argparse
 
 
-def get_loader(dataset, batch_size=64, split=(4, 1)):
+def get_loader(dataset, batch_size=8, split=(4, 1)):
     val_q = len(dataset) // sum(split) * split[1]
 
     val_indices = np.random.choice(list(range(len(dataset))), val_q)
```
Comment (grib0ed0v) [inline src/train.py:16]: Each time different validation dataset?

File: src/datasets.py [hunk 2]
```diff
@@ -22,7 +23,7 @@ def __init__(self, attack_dir, real_dir):
         self.images = self.images + [{'image': os.path.join(real_dir, img),
                                       'label': 'live'}
                                      for img in os.listdir(real_dir)]
-        self.transform = Compose([Resize((480, 640)), ToTensor()])
+        self.transform = Compose([Resize((320, 240)), ToTensor()])
```
Comment (grib0ed0v) [inline src/datasets.py]: I think it's better to have a separate 'transform' function as parameter and just receive it outside of MfsdDataset (in case if you want to experiment with data augmentations)

File: src/datasets.py [hunk 3]
```diff
@@ -22,7 +23,7 @@ def __init__(self, attack_dir, real_dir):
...
 
     def __len__(self):
         return len(os.listdir(self.attack_dir)) +\
```
Comment (grib0ed0v) [inline src/datasets.py]: I don't think it's good idea to ask Operating System size of dir each time when __len__ is called. Maybe you can derive this from **self.images** ?

File: src/datasets.py [hunk 4]
```diff
@@ -22,7 +23,7 @@ def __init__(self, attack_dir, real_dir):
         self.images = self.images + [{'image': os.path.join(real_dir, img),
```
Comment (grib0ed0v) [inline src/datasets.py]: += can't be used?

File: src/datasets.py [hunk 5]
```diff
@@ -40,16 +41,23 @@ def __getitem__(self, idx):
...
+                self.items = [tuple(line[:-1].split(' ')) for line in lines]
+            else:
+                self.items = []
+                for line in lines:
+                    dir_name = os.path.join(line[:-1], 'profile')
+                    for file_name in os.listdir(os.path.join(dir, dir_name)):
+                        item = os.path.join(dir_name, file_name)
+                        self.items.append((item, -1))
+
+        self.transform = Compose([Resize((320, 240)), ToTensor()])
```
Comment (grib0ed0v) [inline src/datasets.py]: The same comment.
self.transform = transform (transform is coming as parameter to constructor)

File: src/submission.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+dev/003000 0.50000dev/003001 0.50000
```
Comment (grib0ed0v) [inline src/submission.txt:1]: Do we really need to track this file in git?

File: src/submit.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                                ......
+                dev/003599 0.23394    #Note:  line 600- the last row of 4@3_dev_res.txt
+    '''
+    argparser = argparse.ArgumentParser()
+    argparser.add_argument('--model1_path', type=str, required=True)
+    argparser.add_argument('--model2_path', type=str, required=True)
+    argparser.add_argument('--model3_path', type=str, required=True)
+    args = argparser.parse_args()
+
+    model = models.mobilenet_v2(num_classes=2)
```
Comment (grib0ed0v) [inline src/submit.py]: it's better to be 2 a parameter to ArgumentParser (--num_classes with default value of 2)

File: src/train.py [hunk 8]
```diff
@@ -1,20 +1,66 @@
...
+    argparser.add_argument('--epochs', type=int, default=10)
+    argparser.add_argument('--checkpoint', type=str)
+    argparser.add_argument('--batch_size', type=int, required=True)
+    argparser.add_argument('--eval_every', type=int, default=1)
+    args = argparser.parse_args()
+    dataset = CasiaSurfDataset(args.protocol)
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    train_loader, val_loader = get_loader(dataset)
+    model = models.mobilenet_v2(
+        num_classes=2) if not args.checkpoint else torch.load(args.checkpoint)
```
Comment (grib0ed0v) [inline src/train.py]: --num_classes 2

File: src/train.py [hunk 9]
```diff
@@ -1,20 +1,66 @@
...
+    model = model.to(device)
+    optimizer = optim.Adam(model.parameters())
```
Comment (grib0ed0v) [inline src/train.py]: learning rate should be a parameter to argparser

File: src/train.py [hunk 10]
```diff
@@ -1,20 +1,66 @@
...
+            val_loss = []
+            with torch.no_grad():
+                for i, batch in enumerate(val_loader):
+                    images, labels = batch
+                    labels = torch.LongTensor(labels)
+                    images, labels = images.to(device), labels.to(device)
+                    outputs = model(images)
+                    loss = loss_fn(outputs, labels)
+                    val_loss.append(loss.item())
+                print(f"\t\t\tValidation loss: {np.mean(val_loss)}")
```
Comment (grib0ed0v) [inline src/train.py]: I think in validation stage we also interested in other metrics? (e.g. accuracy?)

File: src/datasets.py [hunk 11]
```diff
@@ -3,6 +3,7 @@
 from torchvision.transforms import Resize, ToTensor, Compose
```
Comment (grib0ed0v) [inline src/datasets.py]: `import torchvision.transforms as T`, and use everywhere `T.Resize, T.ToTensor, T.Compose`, etc.

**Comment by grib0ed0v** [comment]
@trushkin-ilya let's also add a tensorboard logging. Monitoring graphs is better than analysing logs in terminal.

File: src/train.py (hunk 1 above)
Comment (trushkin-ilya) [reply src/train.py:16]: Yes, in this case validation loss metric won't be the same on different data. On the other hand, using new data in training can improve model generalization ability.

File: src/datasets.py (hunk 2 above)
Comment (trushkin-ilya) [reply src/datasets.py]: There is a must-have `ToTensor()` transform, which converts `PIL.Image` to `torch.Tensor` for loss calculation. So if transforms would be acquired from constructor, at any time you have to specify `ToTensor()` transform to make things work.

File: src/train.py (hunk 9 above)
Comment (trushkin-ilya) [reply src/train.py]: I agree that we should give an ability to specify this somewhere outside. Isn't config file a better place for a model parameter?

File: src/submit.py (hunk 7 above)
Comment (trushkin-ilya) [reply src/submit.py]: I agree that we should give an ability to specify this somewhere outside. Isn't config file a better place for a model parameter?

File: src/datasets.py (hunk 11 above)
Comment (trushkin-ilya) [reply src/datasets.py]: What's the reason of that change? `T` could mean `torch` or something else. Couldn't it be confusing?

File: src/datasets.py (hunk 5 above)
Comment (trushkin-ilya) [reply src/datasets.py]: There is a must-have ToTensor() transform, which converts PIL.Image to torch.Tensor for loss calculation. So if transforms would be acquired from constructor, at any time you have to specify ToTensor() transform to make things work.

File: src/datasets.py (hunk 11 above)
Comment (grib0ed0v) [reply src/datasets.py]: Don't be so boring - `T` was just for example. The idea is to import all transforms as object and then use them for augmentations refer to single object. In contrast if you decide to use something like `RandomHorizontalFlip` later you need to rewrite import once again.

File: src/train.py (hunk 1 above)
Comment (grib0ed0v) [reply src/train.py:16]: The comment was to show that you need to fix random seed, because in this case if you change the model and have an improvement you are not 100% sure, that it comes from this change (not lucky validation subset).

**Comment by trushkin-ilya** [comment]
@grib0ed0v Do you accept the changes?

**Comment by grib0ed0v** [comment]
@trushkin-ilya LGTM!

**Review by grib0ed0v — APPROVED** [review]
